### PR TITLE
feat(quantic): SFINT-3956 Quantic didYouMean component

### PR DESCRIPTION
**Pull request by aelfaleh**
[SFINT-3956](https://coveord.atlassian.net/browse/SFINT-3956) Implement Quantic DidYouMean component

The `quantic-did-you-mean` component is responsible for handling query corrections.

**Corrected automatically**
image

**Corrected manually**
image

### CODE REVIEW

File: packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        this.hasQueryCorrection = this.didYouMean.state.hasQueryCorrection;
+        this.wasAutomaticallyCorrected = this.didYouMean.state.wasAutomaticallyCorrected;
+    }
+
+    applyCorrection() {
+        this.didYouMean.applyCorrection();
+    }
+
+    get noResults() {
+        return this.didYouMean.state.originalQuery;
```
Comment (lbergeron) [inline packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js:60]: I think you should not access the controller state this way, at least not outside the `updateState()` function. I fear your component might become out-of-sync with the controller as LWC refreshes your component based on the variations of the `@track` attributes, not based on the controller state changes per se.

I would suggest to change the `updateState()` function to retrieve all the fields you need from the controller state and store them into attributes marked with `@track`. This way LWC will know which parts of the component should be re-rendered when those attributes change.

File: packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
+import { api, LightningElement, track } from 'lwc';
+import { initializeWithHeadless, registerComponentForInit } from 'c/quanticHeadlessLoader';
+
+export default class QuanticDidYouMean extends LightningElement {
+    @track state = {};
```
Comment (lbergeron) [inline packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js:5]: `state` does not seem to be used.

File: packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    /**
+     * @param {import("coveo").SearchEngine} engine
+     */
+    @api
+    initialize(engine) {
+        this.didYouMean = CoveoHeadless.buildDidYouMean(engine);
+        this.unsubscribe = this.didYouMean.subscribe(() => this.updateState());
+
```
Comment (lbergeron) [inline packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js:51]: You can remove the empty line.

File: packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        this.didYouMean = CoveoHeadless.buildDidYouMean(engine);
+        this.unsubscribe = this.didYouMean.subscribe(() => this.updateState());
+        
+    }
+
+    disconnectedCallback() {
+        if (this.unsubscribe) {
+          this.unsubscribe();
+        }
+      }
```
Comment (lbergeron) [inline packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js:58]: This curly brace is not indented correctly.

**Review by lbergeron — APPROVED** [review]
Looks good to me!

File: packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js (hunk 2 above)
Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js]: Yea if you explode the state like you are doing in the updateState (which is the right way to do it I think), you don't need to track state.

File: packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+    updateState() {
+        this.hasQueryCorrection = this.didYouMean.state.hasQueryCorrection;
+        this.wasAutomaticallyCorrected = this.didYouMean.state.wasAutomaticallyCorrected;
+        this.originalQuery = this.didYouMean.state.originalQuery;
+        this.correctedQuery = this.didYouMean.state.queryCorrection.correctedQuery;
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js:48]: I'm curious, what else is in the `queryCorrection` that we are not using?

**Review by erocheleau — COMMENTED** [review]
Screenshots please :)

File: packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js (hunk 5 above)
Comment (aelfaleh) [reply packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js:48]: `/**
 * Describe correction for a query
 */
export interface QueryCorrection {
  /**
   * The query once corrected
   */
  correctedQuery: string;
  /**
   * Array of correction for each word in the query
   */
  wordCorrections: WordCorrection[];
}

export interface WordCorrection {
  /**
   * Offset, from the beginning of the query
   */
  offset: number;
  /**
   * Length of the correction
   */
  length: number;
  /**
   * The original word that was corrected
   */
  originalWord: string;
  /**
   * The new corrected word
   */
  correctedWord: string;
}`

Comment (aelfaleh) [reply packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js:48]: `queryCorrection` contains:
`correctedQuery` :(string) The query once corrected
`wordCorrections`: (array of  `WordCorrection` ) Array of correction for each word in the query

**Review by btaillon — APPROVED** [review]
Looks good to me as well

**Review by nathanlb — APPROVED** [review]
Other than the tiny tiny indentation issue luc pointed out this looks great! nice work!

**Pull request merged**